clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -318,7 +319,7 @@ public FetchResponse(Struct struct) {
                 if (partitionResponseHeader.hasField(LOG_START_OFFSET_KEY_NAME))
                     logStartOffset = partitionResponseHeader.getLong(LOG_START_OFFSET_KEY_NAME);
 
-                Records records = partitionResponse.getRecords(RECORD_SET_KEY_NAME);
+                T records = (T) partitionResponse.getRecords(RECORD_SET_KEY_NAME);
```

Comment:
The only type that is actually possible here is `MemoryRecords`. I think I would suggest making this constructor private and just using the `MemoryRecords` type explicitly. We can then let `parse` return `FetchResponse<MemoryRecords>`.